bigint-opt initial pull request

td2r opened the pull request.

Commit:
bigint-opt initial commit

kt-cpp commented on bigint-optimized/vector.h:
```diff
@@ -0,0 +1,75 @@
+
+struct vector
+{
+    typedef uint32_t* iterator;
+    typedef uint32_t const* const_iterator;
```
[note] `using`...

td2r replied:
Fixed.

kt-cpp commented on bigint-optimized/vector.h:
```diff
@@ -0,0 +1,75 @@
+    };
+
+    struct buffer_data {
+        size_t capacity;
```
Какие есть причины capacity не находиться рядом с данными?

td2r replied:
Поначалу отверг этот вариант, но сейчас пересмотрел ситуацию. Если мы не боимся более длительного времени доступа к capacity, то логичнее переместить ее к массиву, чем хранить у каждого вектора. Это было единственное, что заставило меня писать так.

td2r replied:
Fixed.

kt-cpp commented on bigint-optimized/vector.h:
```diff
@@ -0,0 +1,75 @@
+    void realloc_if_share_();
+
+private:
+    static const size_t SMALL_SIZE = sizeof(buffer_data) / sizeof(uint32_t);
```
[note] `constexpr`

td2r replied:
Fixed.

kt-cpp commented on bigint-optimized/vector.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+static void unshare(vector::shared_array* a) {
+    if (!--a->ref_counter) {
```
[minor] Лучше использоваться операции сравнения с интегральными данными, а не приведение к `bool`

td2r replied:
Почему?

kt-cpp replied:
Банально так проще читать + часто используется в различных гайдлайнах. Лучше выразить свою мысль более явно, от этого никому не будет хуже, а кому-то даже лучше

td2r replied:
Fixed.

kt-cpp commented on bigint-optimized/vector.cpp:
```diff
@@ -0,0 +1,258 @@
+        // use std::fill instead of memset to activate field static_
+        std::fill(buffers_.static_, buffers_.static_ + size, 0);
+    } else {
+        buffers_.dynamic_ = alloc_data(size);
```
Вообще кажется нет гарантий, что динамическая память будет заполнена 0

td2r replied:
Это так. Fixed.

kt-cpp commented on bigint-optimized/vector.cpp:
```diff
@@ -0,0 +1,258 @@
+    assert(new_capacity > SMALL_SIZE);
+
+    buffer_data new_data = alloc_data(new_capacity);
+    memcpy(new_data.arr->words, buffers_.dynamic_.arr->words, size_ * sizeof(uint32_t));
```
`std::copy`

td2r replied:
Я использовал std::copy, чтобы активировать поле объединения. Здесь это не имеет смысла, поскольку метод вызывается, когда объект большой, а значит dynamic_ уже активирован.

kt-cpp replied:
Несмотря на это `std::copy` в принципе более предпочтительный вариант, т.к. memcpy это наследие С, где не было понятия копирования как в плюсах. К тому же он справляется скомпилиться в столь же эффективный код, что и `memcpy` в зависимости от передаваемого типа

td2r replied:
Fixed.

kt-cpp reviewed (changes requested):
* Есть глобальная проблема, потому опишу тут. Нет смысла при SO сжиматься из big в small. Дорога не сама динамическая память в ваакуме, а её выделение. Если уже выделил память, то нет смысла её понапрасну отдавать. Плюс  в случае когда у тебя часто вставляют и удаляют элементы при граничном размере у тебя буду постоянные аллокации туда сюда
* А ещё кажется имеет смысл вынести больше логики связанной с SO  в соответствующий класс с помощью статических и не очень функций

td2r commented:
- Да, концептуально мне бы самому хотелось так сделать. Однако я узнаю о том, маленький/большой сейчас вектор по его размеру. Размер никак не способен различить две ситуации: когда вектор всегда был маленьким, и когда он из большого стал маленьким. Из-за этого придется хранить флаг, однако тогда моя small-object оптимизация увеличит размер вектора на стеке. Я не знаю, что адекватного могу сделать в такой ситуации.
- Я пытался несколько упростить код через метод data. В resize можно сжать по два случая. Но большего я не вижу, мне нужна еще подсказка.

kt-cpp commented:
* В данном дз вполне можно сохранить флаг и это будет норм. В качестве хака же можно использовать старший бит `size_`, потому что `2 ^ 63 - 1` элементов в коллекции кажется хватит всем на свете
* Ну например всякие функции проверки уникальности, уменьшения и увеличения счётчика, и аллокации памяти можно туда унести

Commit:
bigint-opt fix1
1. the vector no longer narrows
2. small-object logic moved to private and static functions as much as possible
3. typedef iterator replaced with using
4. there is no more buffer_data, capacity placed in shared_array which is shared_buffer now
5. SMALL_SIZE changed to constexpr
6. small fix in unshare: comparing with int instead of cast to bool
7. vector(size_t) fills dynamic memory with 0 values
8. use std::copy instead of memcpy in realloc_data_
9. new function realloc_if_share_or_(cond, new_capacity) which realloc buffer if share or cond and change capacity to new_capacity if and only if cond

td2r commented:
Постарался сделать все как сказано.
- Завел флаг small_
- Перестал сжимать вектор
- Сделал так, что метод, реаллоцирующий буффер может делать это, даже если до вызова слова хранились в статическом буффере
- Вынес немного общего кода в новую функцию.
resize, push_back и pop_back стали заметно короче.

kt-cpp commented on bigint-optimized/vector.cpp:
```diff
@@ -0,0 +1,225 @@
+        std::fill(buffers_.static_, buffers_.static_ + size, 0);
+    } else {
+        buffers_.dynamic_ = alloc_data(size);
+        memset(buffers_.dynamic_->words, 0, size * sizeof(uint32_t));
```
[minor] `std::copy` обсуждали же до этого..

kt-cpp replied:
Их много по коду. Найди все сам

td2r replied:
Но это же не memcpy...
Исправил, но std::memcmp относится к стандарту C++ и std::equal был введен лишь в C++14, поэтому его оставил.

kt-cpp replied:
Справедливо

kt-cpp reviewed (changes requested):
В целом норм кроме твоей любви к сишных функциями вместо плюсовых

Commit:
bigint-opt fix2
memset function replaced with std::fill
